### PR TITLE
Add settings to make training results deterministic

### PR DESCRIPTION
**Pull request by redfler**
# Description

This PR adds a deterministic training path and documentation for Isaac Lab RL workflows.

- Added apps/isaaclab.python.headless.determinism.kit as a deterministic headless rendering experience.
- Updated scripts/reinforcement_learning/rl_games/train.py to add opt-in --deterministic and use configure_seed(env_cfg.seed, args_cli.deterministic).
- Updated docs/source/features/reproducibility.rst to document --experience isaaclab.python.headless.determinism.kit
and clarify that strict PyTorch determinism is currently exposed only for RL-Games.

Test command example:

./isaaclab.sh -p scripts/reinforcement_learning/rl_games/train.py --task Isaac-Cartpole-RGB-v0 --enable_cameras --headless --seed 42 --max_iteration 20 **--deterministic --experience isaaclab.python.headless.determinism.kit**

Fixes # (issue)

https://github.com/isaac-sim/IsaacLab/issues/3505 Non-reproducible training results in vision-based tasks with identical seeds

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## Screenshots

| Before | After |
| ------ | ----- |
| Before | After |

## Checklist

- [x] I have read and understood the [contribution guidelines](https://isaac-sim.github.io/IsaacLab/main/source/refs/contributing.html)
- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

File: apps/isaaclab.python.headless.determinism.kit [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+title = "Isaac Lab Python Headless Deterministic Rendering"
+description = "An app for running Isaac Lab headlessly with rendering enabled and deterministic settings"
+version = "3.0.0"
+
+# That makes it browsable in UI with "experience" filter
+keywords = ["experience", "app", "isaaclab", "python", "camera", "minimal"]
+
+[dependencies]
+# Isaac Lab minimal app
+"isaaclab.python.headless" = {}
```
Comment (kellyguo11) [inline apps/isaaclab.python.headless.determinism.kit]: probably better to depend on the headless rendering app? and then we can avoid the additional dependencies below

File: scripts/reinforcement_learning/rl_games/train.py [hunk 2]
```diff
@@ -48,6 +49,15 @@
     "--agent", type=str, default="rl_games_cfg_entry_point", help="Name of the RL agent configuration entry point."
 )
 parser.add_argument("--seed", type=int, default=None, help="Seed used for the environment")
+parser.add_argument(
+    "--deterministic",
```
Comment (kellyguo11) [inline scripts/reinforcement_learning/rl_games/train.py]: can this flag be added to AppLauncher instead so that we can switch to using the reproducible .kit file automatically?

File: docs/source/features/reproducibility.rst [hunk 3]
```diff
@@ -20,6 +20,33 @@ simulation results are reproducible across different runs. The seed is set into
...
+
+   The optional ``--deterministic`` flag (which calls :meth:`~isaaclab.utils.seed.configure_seed` with
+   ``torch_deterministic=True``, enabling ``torch.use_deterministic_algorithms(True)`` and related CUDNN
+   settings) exists only on ``scripts/reinforcement_learning/rl_games/train.py``.
+
+   The RSL-RL, Stable-Baselines3, SKRL, and RLinf training scripts under
+   ``scripts/reinforcement_learning/`` still honor ``--seed`` / agent configuration for the Isaac Lab
+   environment and learning stack, but they do **not** expose an equivalent opt-in for strict PyTorch-wide
+   deterministic algorithms. If you need that behavior with another framework, call
+   :meth:`~isaaclab.utils.seed.configure_seed` with ``torch_deterministic=True`` from your own training
```
Comment (kellyguo11) [inline docs/source/features/reproducibility.rst]: what's the reason this can't be added to the other training scripts?

File: apps/isaaclab.python.headless.determinism.kit (hunk 1 above)
Comment (redfler) [reply apps/isaaclab.python.headless.determinism.kit]: done

File: scripts/reinforcement_learning/rl_games/train.py (hunk 2 above)
Comment (redfler) [reply scripts/reinforcement_learning/rl_games/train.py]: done

File: docs/source/features/reproducibility.rst (hunk 3 above)
Comment (redfler) [reply docs/source/features/reproducibility.rst]: This PR concerns the Isaac-Cartpole-RGB-v0 task. In the Gymnasium registry, that environment id presently exposes only the RL-Games configuration entry point, and does not include Stable-Baselines3, SKRL, or RSL-RL entry points.
https://github.com/isaac-sim/IsaacLab/blob/develop/source/isaaclab_tasks/isaaclab_tasks/manager_based/classic/cartpole/__init__.py#L38

File: apps/isaaclab.python.headless.determinism.kit [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+# note: this is done to ensure prior versions of Isaac Sim still use the latest assets
+[settings]
+persistent.isaac.asset_root.default = "https://omniverse-content-staging.s3-us-west-2.amazonaws.com/Assets/Isaac/6.0"
+persistent.isaac.asset_root.cloud = "https://omniverse-content-staging.s3-us-west-2.amazonaws.com/Assets/Isaac/6.0"
+persistent.isaac.asset_root.nvidia = "https://omniverse-content-staging.s3-us-west-2.amazonaws.com/Assets/Isaac/6.0"
+
+# ═══════════════════════════════════════════════════════════
+# DETERMINISM-CRITICAL: changing these may break reproducibility
+# ═══════════════════════════════════════════════════════════
+rtx.rendermode = "RealTimePathTracing"
```
Comment (kellyguo11) [inline apps/isaaclab.python.headless.determinism.kit]: do these settings also work with enabling determinism in non-headless mode? is there any reason this .kit file needs to be limited to  headless? what if we can enable these at runtime if users ask for determinism, so that it doesn't really matter which .kit file ends up getting used?

File: docs/source/features/reproducibility.rst [hunk 5]
```diff
@@ -20,6 +20,83 @@ simulation results are reproducible across different runs. The seed is set into
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``--deterministic`` flag is provided by :meth:`isaaclab.app.AppLauncher.add_app_launcher_args`.
+When used with the default experience selection logic in a compatible headless launch, AppLauncher
+automatically selects ``isaaclab.python.headless.determinism.kit``.
+
+**Strict PyTorch determinism** (calling :meth:`~isaaclab.utils.seed.configure_seed` with
+``torch_deterministic=True``) is wired into the **RL-Games** training script only. Other frameworks
+still honor ``--seed`` / agent configuration for the environment; use
+:meth:`~isaaclab.utils.seed.configure_seed` from your own entry point if you need the same PyTorch-wide
```
Comment (kellyguo11) [inline docs/source/features/reproducibility.rst]: if users pass --deterministic, I think it'd also make sense to force this call in train.py / play.py scripts in the other RL library scripts?

File: docs/source/features/reproducibility.rst [hunk 6]
```diff
@@ -20,6 +20,83 @@ simulation results are reproducible across different runs. The seed is set into
...
+behavior elsewhere.
+
+To enable deterministic rendering/app settings, launch workflows with ``--deterministic``
+**and** ``--enable_cameras`` **and** ``--headless`` (without livestream/XR):
```
Comment (kellyguo11) [inline docs/source/features/reproducibility.rst]: both `--enable_cameras` and `--headless` should no longer be necessary

**Review by kellyguo11 — COMMENTED** [review]
My feeling is we don't need to tie the deterministic flag towards a specific RL library, especially since rl-games is less maintained now and we are shifting away from it. Could we do the same setup that enables determinism for RL Games in the train and play scripts for other RL libraries as well?

**Comment by kellyguo11** [comment]
@StafaH for viz, this may impact your training script refactoring

File: apps/isaaclab.python.headless.determinism.kit (hunk 4 above)
Comment (redfler) [reply apps/isaaclab.python.headless.determinism.kit]: enable determinism at runtime

File: docs/source/features/reproducibility.rst (hunk 5 above)
Comment (redfler) [reply docs/source/features/reproducibility.rst]: As for task Isaac-Cartpole-v0, it is deterministic, no need to use deterministic flag and call configure_seed. 
As for task Isaac-Cartpole-RGB-v0, it is non-deterministic, need to use deterministic flag and call configure_seed.
task Isaac-Cartpole-RGB-v0 is only supported in RL-Games, so not change the other RL training scripts

File: docs/source/features/reproducibility.rst [hunk 7]
```diff
@@ -20,6 +20,82 @@ simulation results are reproducible across different runs. The seed is set into
...
+The ``--deterministic`` flag is provided by :meth:`isaaclab.app.AppLauncher.add_app_launcher_args`.
+After the simulation app starts, :class:`~isaaclab.app.app_launcher.AppLauncher` applies RTX/RTPT carb
+settings via :meth:`~isaaclab.app.app_launcher.AppLauncher.apply_rtx_determinism_settings`.
+
+**Strict PyTorch determinism** (calling :meth:`~isaaclab.utils.seed.configure_seed` with
+``torch_deterministic=True`` when you pass ``--deterministic``) is wired into the RL training scripts
+for **RL-Games**, **skrl**, **RSL-RL**, and **Stable-Baselines3**: each calls
+:meth:`~isaaclab.utils.seed.configure_seed` after constructing its framework runner or agent object
+so library initialization is not disturbed, then training proceeds with the requested global RNG and
+optional PyTorch deterministic algorithms. As for task Isaac-Cartpole-v0, it is deterministic,
```
Comment (kellyguo11) [inline docs/source/features/reproducibility.rst]: instead of mentioning Isaac-Cartpole-v0 and Isaac-Cartpole-RGB-v0 explicitly, maybe we can generalize this towards:

- general physics-only simulation does not require the flag
- RTX rendering (non-minimal mode) requires the flag
- Newton rendering does not require the flag

File: docs/source/features/reproducibility.rst [hunk 8]
```diff
@@ -20,6 +20,82 @@ simulation results are reproducible across different runs. The seed is set into
...
+so you need to use deterministic flag and call configure_seed.
+
+To enable deterministic RTX settings from the app launcher, pass ``--deterministic``.
+
+.. code-block:: bash
+
+  ./isaaclab.sh -p scripts/reinforcement_learning/rl_games/train.py \
+    --task Isaac-Cartpole-RGB-v0 --enable_cameras --headless --deterministic
+
+Gymnasium registry (``gym.register``) and training scripts
```
Comment (kellyguo11) [inline docs/source/features/reproducibility.rst]: this section doesn't really feel necessary for the reproducibility docs

File: docs/source/features/reproducibility.rst [hunk 9]
```diff
@@ -20,6 +20,82 @@ simulation results are reproducible across different runs. The seed is set into
...
+  optional; **only keys that appear in ``kwargs`` are valid** for that task id.
+
+When you run ``scripts/reinforcement_learning/<framework>/train.py --task <TASK_ID>``, the script resolves
+``<framework>_cfg_entry_point`` (or the name you pass with ``--agent``) against the registry. If the task
+was registered **without** that key—for example ``Isaac-Cartpole-RGB-v0`` currently lists only
+``rl_games_cfg_entry_point``—you will get a ``ValueError`` such as “Could not find configuration …
+``sb3_cfg_entry_point``”. To use another framework you must either pick a task that registers that entry
+point or extend ``gym.register(..., kwargs={...})`` for that task with matching agent YAML/Python configs.
+
+Regression test for training scripts
```
Comment (kellyguo11) [inline docs/source/features/reproducibility.rst]: I don't think we need to document regression tests for users. we can explain this test as a script that can be used to check for determinism instead of calling it regression test

File: scripts/reinforcement_learning/rl_games/train.py [hunk 10]
```diff
@@ -208,6 +209,10 @@ def main():
         else:
             runner = Runner(IsaacAlgoObserver())
 
+        # configure_seed must be called after Runner() so that PyTorch deterministic settings
+        # do not interfere with Runner's internal initialization.
+        configure_seed(env_cfg.seed, args_cli.deterministic)
```
Comment (kellyguo11) [inline scripts/reinforcement_learning/rl_games/train.py]: should we include this option for play.py as well?

for the other libraries, we can maybe try a different environment to test. `Isaac-Repose-Cube-Shadow-Vision-Direct-v0` for example supports RTX rendering with RSL RL.

otherwise, I think it'll be good to include this call anyways for all the library scripts for consistency. as long as the script still works for existing environments, it should be fine. I don't think this call would be harmful.

**Review by kellyguo11 — APPROVED** [review]
